File: java-gilded_rose-20160427/src/test/java/com/gildedrose/GildedRoseTest.java
package com.gildedrose;

import static org.junit.Assert.*;

import org.junit.Test;

public class GildedRoseTest {

    @Test
    public void updateQuality_appIsInitializedWithItems() {
        Item[] items = new Item[]{new Item("foo", 0, 0)};
        GildedRose app = new GildedRose(items);

        app.updateQuality();

        assertEquals("foo", app.items[0].name);
    }

    @Test
    public void updateQuality_decreasesSellin() {
        Item[] items = new Item[]{new Item("foo", 2, 2)};
        GildedRose app = new GildedRose(items);

        app.updateQuality();

        assertEquals(1, app.items[0].sellIn);
    }

    @Test
    public void updateQuality_decreasesQuality() {
        Item[] items = new Item[]{new Item("foo", 2, 2)};
        GildedRose app = new GildedRose(items);

        app.updateQuality();

        assertEquals(1, app.items[0].quality);
    }


    @Test
    public void updateQuality_degradesTwiceFasterWhenSellInIsZero() {
        Item[] items = new Item[]{new Item("foo", 1, 20)};
        GildedRose app = new GildedRose(items);

        app.updateQuality();

        assertEquals(19, app.items[0].quality);

        app.updateQuality();

        assertEquals(17, app.items[0].quality);
    }

    @Test
    public void updateQuality_qualityNeverNegative() {
        Item[] items = new Item[]{new Item("foo", 0, 0)};
        GildedRose app = new GildedRose(items);

        app.updateQuality();

        assertEquals(0, app.items[0].quality);
    }

    @Test
    public void updateQuality_agedBriedIncreasesQuality() {
        Item[] items = new Item[]{new Item("Aged Brie", 10, 10)};
        GildedRose app = new GildedRose(items);

        app.updateQuality();

        assertEquals(11, app.items[0].quality);
    }

    @Test
    public void updateQuality_qualityNeverHigherThan50() {
        Item[] items = new Item[]{new Item("Aged Brie", 10, 50)};
        GildedRose app = new GildedRose(items);

        app.updateQuality();

        assertEquals(50, app.items[0].quality);
    }

    @Test
    public void updateQuality_sulfurasNeverChange() {
        Item[] items = new Item[]{new Item("Sulfuras, Hand of Ragnaros", 10, 20)};
        GildedRose app = new GildedRose(items);

        app.updateQuality();

        assertEquals(10, app.items[0].sellIn);
        assertEquals(20, app.items[0].quality);
    }

    @Test
    public void updateQuality_backstageIncreasesQualityBy2WhenSellInIs10orLess() {
        Item[] items = new Item[]{
                new Item("Backstage passes to a TAFKAL80ETC concert", 10, 10),
                new Item("Backstage passes to a TAFKAL80ETC concert", 6, 10)
        };
        GildedRose app = new GildedRose(items);

        app.updateQuality();

        assertEquals(12, app.items[0].quality);
        assertEquals(12, app.items[1].quality);
    }

    @Test
    public void updateQuality_backstageIncreasesQualityBy3WhenSellInIs5orLess() {
        Item[] items = new Item[]{
                new Item("Backstage passes to a TAFKAL80ETC concert", 5, 10),
                new Item("Backstage passes to a TAFKAL80ETC concert", 1, 10)
        };
        GildedRose app = new GildedRose(items);

        app.updateQuality();

        assertEquals(13, app.items[0].quality);
        assertEquals(13, app.items[1].quality);
    }

    @Test
    public void updateQuality_backstageDropsToZeroWhenSellInIsZero() {
        Item[] items = new Item[]{
                new Item("Backstage passes to a TAFKAL80ETC concert", 0, 10)};
        GildedRose app = new GildedRose(items);

        app.updateQuality();

        assertEquals(0, app.items[0].quality);
    }

    @Test
    public void updateQuality_conjuredDegradesTwiceAsNormal() {
        Item[] items = new Item[]{
                new Item("Conjured", 10, 10),
                new Item("Conjured", 0, 10)
        };
        GildedRose app = new GildedRose(items);

        app.updateQuality();

        assertEquals(8, app.items[0].quality);
        assertEquals(6, app.items[1].quality);
    }

}

File: 20191009-kotlin-gilded-rose/src/test/kotlin/com/gildedrose/GildedRoseTest.kt
package com.gildedrose

import org.assertj.core.api.Assertions.assertThat
import org.junit.Test

class GildedRoseTest {

    @Test
    fun `quality and sellIn decreases by 1`() {
        val items = arrayOf(Item("An Item", 1, 10))
        val gilgedRose = GildedRose(items)

        gilgedRose.updateQuality()

        assertThat(items[0].sellIn).isEqualTo(0)
        assertThat(items[0].quality).isEqualTo(9)
    }

    @Test
    fun `once the sell by date has passed quality decreases twice as fast`() {
        val items = arrayOf(Item("An Item", 0, 10))
        val gilgedRose = GildedRose(items)

        gilgedRose.updateQuality()

        assertThat(items[0].quality).isEqualTo(8)
    }

    @Test
    fun `the quality of an item is never negative`() {
        val items = arrayOf(Item("An Item", 0, 0))
        val gilgedRose = GildedRose(items)

        gilgedRose.updateQuality()

        assertThat(items[0].quality).isEqualTo(0)
    }

    @Test
    fun `aged brie increases in quality the older it gets`() {
        val items = arrayOf(Item("Aged Brie", 1, 10))
        val gilgedRose = GildedRose(items)

        gilgedRose.updateQuality()

        assertThat(items[0].quality).isEqualTo(11)
    }

    @Test
    fun `aged brie when sellIn expires`() {
        val items = arrayOf(Item("Aged Brie", 0, 10))
        val gilgedRose = GildedRose(items)

        gilgedRose.updateQuality()

        assertThat(items[0].sellIn).isEqualTo(-1)
        assertThat(items[0].quality).isEqualTo(12)
    }

    @Test
    fun `the quality of an item is never more than 50`() {
        val items = arrayOf(Item("Aged Brie", 1, 50))
        val gilgedRose = GildedRose(items)

        gilgedRose.updateQuality()

        assertThat(items[0].quality).isEqualTo(50)
    }

    @Test
    fun `sulfuras never has to be sold or decreases in quality`() {
        val items = arrayOf(Item("Sulfuras, Hand of Ragnaros", 1, 80))
        val gilgedRose = GildedRose(items)

        gilgedRose.updateQuality()

        assertThat(items[0].quality).isEqualTo(80)
    }

    @Test
    fun `backstage passes increases in quality by 2 when there are 10 days or less`() {
        val items = arrayOf(Item("Backstage passes to a TAFKAL80ETC concert", 10, 10))
        val gilgedRose = GildedRose(items)

        gilgedRose.updateQuality()

        assertThat(items[0].quality).isEqualTo(12)
    }

    @Test
    fun `backstage passes increases in quality by 3 when there are 5 days or less`() {
        val items = arrayOf(Item("Backstage passes to a TAFKAL80ETC concert", 5, 10))
        val gilgedRose = GildedRose(items)

        gilgedRose.updateQuality()

        assertThat(items[0].quality).isEqualTo(13)
    }

    @Test
    fun `backstage passes drops quality to 0 after concert`() {
        val items = arrayOf(Item("Backstage passes to a TAFKAL80ETC concert", 0, 10))
        val gilgedRose = GildedRose(items)

        gilgedRose.updateQuality()

        assertThat(items[0].quality).isEqualTo(0)
    }
}
